internal/action/generate.go

```diff
@@ -261,21 +261,40 @@ func clamp(mi, ma, value int) int {
 	return value
 }
 
-func (s *Action) generatePasswordForRule(ctx context.Context, c *cli.Context, length, name, domain string, rule pwrules.Rule) (string, error) {
+// generatePasswordForRule validates the user-provided password length against
+// the rule for the domain and condtionally prompts the user for a correct
+// length if the initial value is invalid.
+func (s *Action) generatePasswordForRule(ctx context.Context, length, domain string, rule pwrules.Rule) (string, error) {
 	out.Noticef(ctx, "Using password rules for %s ...", domain)
 
-	wl := 16
-	if iv, err := strconv.Atoi(length); err == nil {
-		wl = clamp(rule.Minlen, rule.Maxlen, iv)
-		debug.Log("restricting requested password length (%s) to %d because of the rule {%d,%d}", length, wl, rule.Minlen, rule.Maxlen)
-	}
+	var iv int
+	var err error
 
-	question := fmt.Sprintf("How long should the password be? (min: %d, max: %d)", rule.Minlen, rule.Maxlen)
-	iv, err := termio.AskForInt(ctx, question, wl)
-	if err != nil {
+	if iv, err = strconv.Atoi(length); err != nil {
 		return "", exit.Error(exit.Usage, err, "password length must be a number")
 	}
 
+	wl := clamp(rule.Minlen, rule.Maxlen, iv)
+	if wl != iv {
+		debug.Log(
+			"setting requested password length (%s) to %d because of the rule {min: %d, max: %d}",
+			length, wl, rule.Minlen, rule.Maxlen,
+		)
+
+		question := fmt.Sprintf(
+			"How long should the password be? (min: %d, max: %d)",
+			rule.Minlen, rule.Maxlen,
+		)
+
+		var sv string
+
+		if sv, err = termio.AskForString(ctx, question, strconv.Itoa(wl)); err != nil {
+			return "", err
+		}
+
+		return s.generatePasswordForRule(ctx, sv, domain, rule)
```

Comment:
That's a recursive call? Why can't you just carry on with the rest of the function?
Is it in order to prevent the user from inputting an invalid length, so you'd just keep asking if they do so?

Reply:
yes. see the initial discussion: https://github.com/gopasspw/gopass/pull/3159#discussion_r2100552140

Reply:
I think it warrants a little comment about why it's recursive in the code itself for future maintainer working on this not to have to dig commits / PRs.

Reply:
i described it in the commit body (`git-blame`, `git-log -L`) but can definitely add a comment just before this return.

Reply:
i added a comment above this statement in d9364941312feebadaba7d6b91c9df98c82bce49.